tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -12930,6 +12930,40 @@ bool ValidateAttributeTargetIsFunction(Sema& S, Decl* D, const AttributeList &A)
   return false;
 }
 
+void Sema::ValidateShaderAttributes(Decl *D, const AttributeList *A) {
```

Comment:
I think you can refactor this code to be much simpler.

The code just above here is actually creating the AST node for the attribute. This block should get hit once for each time the `[shader(...)]` block shows up in the source. Instead of iterating over the list of attributes, you can rely on this being hit over and over again.

If you validate the attribute before creating the AST node (so you don't create the node if it is invalid), you can use the existing AST node as the "lastAttr" value.

That could simplify this code to something more like:

```c++
HLSLShaderAttr *ValidateShaderAttributes(Sema &S, Decl *D, const AttributeList *A) {
  Expr *ArgExpr = A->getArgAsExpr(0);
  StringLiteral *Literal = dyn_cast<StringLiteral>(ArgExpr->IgnoreParenCasts());
  DXIL::ShaderKind Stage = ShaderModel::KindFromFullName(Literal->getString());
  if (Stage == DXIL::ShaderKind::Invalid) {
    S.Diag(...); // Diagnose invalid stage specifier.
    return nullptr; // don't create the attribute
  }
  HLSLShaderAttr *Existing = D->getAttr<HLSLShaderAttr>();
  if (Existing) {
    DXIL::ShaderKind NewStage = ShaderModel::KindFromFullName(Existing->getStage());
   if (Stage == NewStage)
    return nullptr; // don't create, but no error.
    if ((Stage != DXIL::ShaderKind::Compute && Stage != DXIL::ShaderKind::Node) ||
        (NewStage != DXIL::ShaderKind::Compute && NewStage != DXIL::ShaderKind::Node)) {
      S.Diag(A->getLoc(), diag::err_hlsl_attribute_mismatch);
      S.Diag(A->getLoc(), diag::note_conflicting_attribute);
      S.Diag(Existing->getLoc(), diag::note_conflicting_attribute);
      return nullptr;
    }
  }
  return ::new (S.Context) HLSLShaderAttr(
        A.getRange(), S.Context,Literal->getString())
}
```

The idea here is that the function gets called each time there is a parsed `shader` attribute, but it only creates the AST node if they are valid. This would allow something like:

```hlsl
[shader("compute")]
[shader("node")]
[shader("compute")]
[shader("node")]
```

But it should fail for combinations of other shaders. We could additionally de-dupe compute/node shader attributes, but we don't do that today so I don't know that we need to.

Reply:
- I think my implementation is sufficiently simple.
- With the way the code is structured, there is no way to gain access to the source location of the previous attribute. My implementation allows access to an accurate source location for the previous attribute because of the usage of a loop. The Decl variable D has no relevant source location to the new attribute being processed, and "Existing" has no source location available either.
- This implementation differs from every other validation function in that it takes responsibility for creating the new attribute object. Because of this, I need to personally set Handled to true if nullptr is returned, and it seems to defeat the purpose of the switch statement that calls this validation function. To match the convention with the rest of the validation in the switch statement I don't think this function should create/return the Attr object.
- My implementation is correct when testing the CNCN case you pasted above.

Unless there's further reason, I think I should leave the code as is.

Reply:
The implementation I proposed is how clang does attribute verification. The HLSL pattern is not really better.
I'm unsure why the `Existing` attribute in my example wouldn't have a valid source location. It should. If it doesn't that would be a bug.